File: LaddersAndSlides_GameEngine/Engine/GameEngine.cs
namespace LaddersAndSlides_GameEngine.Engine
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Domain;
    using LaddersAndSlidesW8UI.Processing;
    using Windows.UI.Xaml.Controls;
    using Windows.UI.Xaml.Media;
    using Windows.UI.Xaml.Media.Imaging;

    public class GameEngine
    {
        public List<Player> Players { get; set; }
        public List<SpecialMove> SpecialMoves { get; set; } 
        public Player CurrentPlayer { get; set; }
        public GameStateEngine CurrentState { get; set; }
        public SpecialMove CurrentSpecialMove { get; set; }

        protected Random RandomNumberGenerator { get; set; }
        protected double ArrowSpinDuration { get; set; }
        protected int NumberOfMovesForCurrentPlayer { get; set; }
        protected DateTime WaitStart { get; set; }

        public GameEngine()
        {
            CreateListSpecialMoves();
        }

        public void GetNextPlayer()
        {
            if (CurrentPlayer == null)
            {
                CurrentPlayer = Players.FirstOrDefault(x => x.TurnOrder == 1);
            }
            else
            {
                var nextTurnOrder = CurrentPlayer.TurnOrder + 1;

                CurrentPlayer = nextTurnOrder > Players.Count 
                    ? Players.FirstOrDefault(x => x.TurnOrder == 1) 
                    : Players.FirstOrDefault(x => x.TurnOrder == nextTurnOrder);
            }

            CurrentState = GameStateEngine.TurnComplete;
        }

        public void ProcessArrowEvent(Image arrow)
        {
            arrow.Projection = new PlaneProjection { CenterOfRotationY = 0.5, CenterOfRotationX = 0.5, RotationZ = ArrowSpinDuration };
            ArrowSpinDuration *= .98;

            if (RandomNumberGenerator.NextDouble() < .05 && ArrowSpinDuration < 360)
            {
                CurrentState = GameStateEngine.ArrowDelayedEvent;
                NumberOfMovesForCurrentPlayer = CalculateNumberOfPlayerMoves(ArrowSpinDuration);
                WaitStart = DateTime.Now;
            }
        }

        public Image CreatePlayerToken(Player player)
        {
            var bitmapImage = new BitmapImage { UriSource = player.ImageUri };

            var image = new Image
            {
                Width = 55,
                Height = 55,
                Name = player.Name,
                Source = bitmapImage,
                RenderTransform = new CompositeTransform()
            };

            return image;
        }

        public int CalculateNumberOfPlayerMoves(double arrowSpinAngle)
        {
            if ((arrowSpinAngle >= 0 && arrowSpinAngle <= 40) ||
                (arrowSpinAngle > 320 && arrowSpinAngle <= 360))
                return 4;

            if (arrowSpinAngle > 40 && arrowSpinAngle <= 95)
                return 3;

            if (arrowSpinAngle > 95 && arrowSpinAngle <= 150)
                return 2;

            if (arrowSpinAngle > 150 && arrowSpinAngle <= 209)
                return 1;

            if (arrowSpinAngle > 209 && arrowSpinAngle <= 260)
                return 6;

            return 5;
        }

        public void CalculateArrowSpin()
        {
            //generate new random number used for calculation
            RandomNumberGenerator = new Random();
            ArrowSpinDuration = RandomNumberGenerator.Next(2500) + RandomNumberGenerator.Next(200, 800);
        }

        public void ProcessArrowDelayEvent()
        {
            if (DateTime.Now.Subtract(WaitStart).Seconds > .45) // Wait a half second before starting the movement of the player token.
            {
                CurrentState = GameStateEngine.PlayerEvent;
            }
        }

        public void ProcessPlayerEvent(Image currentPlayerToken, Grid gutter, Grid gameBoard)
        {
            if (DateTime.Now.Subtract(WaitStart).Milliseconds > 100)
            {
                WaitStart = DateTime.Now;

                if (!CurrentPlayer.TurnInProcess)
                {
                    CurrentPlayer.CurrentMovesRemaining = NumberOfMovesForCurrentPlayer;
                    CurrentPlayer.TurnInProcess = true;
                }

                if (currentPlayerToken.Parent == gutter)
                {
                    gutter.Children.Remove(currentPlayerToken);
                    gameBoard.Children.Add(currentPlayerToken);
                    
                    Grid.SetRow(currentPlayerToken, 9);
                    Grid.SetColumn(currentPlayerToken, 0);
                    
                    CurrentPlayer.CurrentMovesRemaining--;
                    CurrentPlayer.CurrentTileNumber = 1;
                    CurrentPlayer.CurrentGridColumn = 0;
                    CurrentPlayer.CurrentGridRow = 9;
                }
                else
                {
                    var currentColumn = Grid.GetColumn(currentPlayerToken);
                    var currentRow = Grid.GetRow(currentPlayerToken);

                    if (CurrentPlayer.CurrentTileNumber + CurrentPlayer.CurrentMovesRemaining > 100)
                    {
                        CurrentPlayer.CurrentMovesRemaining = 0;
                    }
                    else
                    {
                        MakePlayerMove(currentPlayerToken, currentRow, currentColumn);
                    }
                }

                if (CurrentPlayer.CurrentMovesRemaining <= 0)
                {
                    CurrentState = GameStateEngine.GetNextPlayer;
                    CurrentPlayer.TurnInProcess = false;

                    var specialMove = IsSpecialMove(CurrentPlayer.CurrentTileNumber);
                    
                    if (specialMove)
                    {
                        CurrentState = GameStateEngine.PlayerSpecialMoveTransportMoveEvent;
                    }
                    else if (CurrentPlayer.CurrentTileNumber == 100)
                    {
                        CurrentState = GameStateEngine.WinnerDeclared;
                    }
                }
            }
        }

        internal void MakePlayerMove(Image currentPlayerToken, int currentRow, int currentColumn)
        {
            if (CurrentPlayer.CurrentTileNumber%10 == 0)
            {
                Grid.SetRow(currentPlayerToken, currentRow - 1);
                CurrentPlayer.CurrentlyOnAlternateRow = currentRow%2 != 0;
            }
            else
            {
                if (CurrentPlayer.CurrentlyOnAlternateRow)
                {
                    Grid.SetColumn(currentPlayerToken, currentColumn - 1);
                }
                else
                {
                    Grid.SetColumn(currentPlayerToken, currentColumn + 1);
                }
            }

            CurrentPlayer.CurrentTileNumber++;
            CurrentPlayer.CurrentMovesRemaining--;
            CurrentPlayer.CurrentGridColumn = Grid.GetColumn(currentPlayerToken);
            CurrentPlayer.CurrentGridRow = Grid.GetRow(currentPlayerToken);
        }

        public void MakeSpecialMove(Image currentPlayerToken, Grid gameBoard)
        {
            Grid.SetColumn(currentPlayerToken, CurrentSpecialMove.EndingGridColumn);
            Grid.SetRow(currentPlayerToken, CurrentSpecialMove.EndingGridRow);
            CurrentPlayer.CurrentTileNumber = CurrentSpecialMove.EndingTileNumber;
            CurrentPlayer.CurrentlyOnAlternateRow = CurrentSpecialMove.EndingGridRow % 2 == 0;
            CurrentState = GameStateEngine.GetNextPlayer;
            CurrentPlayer.TurnInProcess = false;
        }

        internal bool IsSpecialMove(int playerTile)
        {
            var result = SpecialMoves.FirstOrDefault(x => x.StartingTileNumber == playerTile);

            if (result != null)
            {
                CurrentSpecialMove = result;
                CurrentPlayer.SpecialMoveTransportDestination = result.EndingTileNumber;
            }

            return result != null;
        }

        public void CreateListSpecialMoves()
        {
            SpecialMoves = new List<SpecialMove>();

            var specialMoves = new Dictionary<int, int>
                {
                    {1, 38}, {4, 14}, {9, 31}, {21, 42},
                    {28, 84}, {51, 67}, {71, 91}, {80, 100},
                    {16,2}, {47,26}, {49,11}, {56,53},
                    {64,60}, {87,24}, {93,73}, {95,75},
                    {98,78}
                };

            foreach (var specialMove in specialMoves)
            {
                Dictionary<string, int> startingGridLocation = GetGridColumnRow(specialMove.Key);
                Dictionary<string, int> endingGridLocation = GetGridColumnRow(specialMove.Value);

                var sMove = new SpecialMove
                    {
                        StartingTileNumber = specialMove.Key,
                        StartingGridColumn = startingGridLocation["column"],
                        StartingGridRow = startingGridLocation["row"],
                        EndingTileNumber = specialMove.Value,
                        EndingGridColumn = endingGridLocation["column"],
                        EndingGridRow = endingGridLocation["row"]
                    };

                SpecialMoves.Add(sMove);
            }
        }

        private Dictionary<string, int> GetGridColumnRow(int tileNumber)
        {
            var gridLocation = new Dictionary<string, int>();

            switch (tileNumber)
            {
                case 1:
                    gridLocation.Add("column", 0);
                    gridLocation.Add("row", 9);
                    break;
                case 2:
                    gridLocation.Add("column", 1);
                    gridLocation.Add("row", 9);
                    break;
                case 4:
                    gridLocation.Add("column", 3);
                    gridLocation.Add("row", 9);
                    break;
                case 9:
                    gridLocation.Add("column", 8);
                    gridLocation.Add("row", 9);
                    break;
                case 11:
                    gridLocation.Add("column", 9);
                    gridLocation.Add("row", 8);
                    break;
                case 14:
                    gridLocation.Add("column", 6);
                    gridLocation.Add("row", 8);
                    break;
                case 16:
                    gridLocation.Add("column", 4);
                    gridLocation.Add("row", 8);
                    break;
                case 21:
                    gridLocation.Add("column", 0);
                    gridLocation.Add("row", 7);
                    break;
                case 24:
                    gridLocation.Add("column", 3);
                    gridLocation.Add("row", 7);
                    break;
                case 26:
                    gridLocation.Add("column", 5);
                    gridLocation.Add("row", 7);
                    break;
                case 28:
                    gridLocation.Add("column", 7);
                    gridLocation.Add("row", 7);
                    break;
                case 31:
                    gridLocation.Add("column", 9);
                    gridLocation.Add("row", 6);
                    break;
                case 38:
                    gridLocation.Add("column", 2);
                    gridLocation.Add("row", 6);
                    break;
                case 42:
                    gridLocation.Add("column", 1);
                    gridLocation.Add("row", 5);
                    break;
                case 47:
                    gridLocation.Add("column", 6);
                    gridLocation.Add("row", 5);
                    break;
                case 49:
                    gridLocation.Add("column", 8);
                    gridLocation.Add("row", 5);
                    break;
                case 51:
                    gridLocation.Add("column", 9);
                    gridLocation.Add("row", 4);
                    break;
                case 53:
                    gridLocation.Add("column", 7);
                    gridLocation.Add("row", 4);
                    break;
                case 56:
                    gridLocation.Add("column", 4);
                    gridLocation.Add("row", 4);
                    break;
                case 60:
                    gridLocation.Add("column", 0);
                    gridLocation.Add("row", 4);
                    break;
                case 64:
                    gridLocation.Add("column", 3);
                    gridLocation.Add("row", 3);
                    break;
                case 67:
                    gridLocation.Add("column", 6);
                    gridLocation.Add("row", 3);
                    break;
                case 71:
                    gridLocation.Add("column", 9);
                    gridLocation.Add("row", 2);
                    break;
                case 73:
                    gridLocation.Add("column", 7);
                    gridLocation.Add("row", 2);
                    break;
                case 75:
                    gridLocation.Add("column", 5);
                    gridLocation.Add("row", 2);
                    break;
                case 78:
                    gridLocation.Add("column", 2);
                    gridLocation.Add("row", 2);
                    break;
                case 80:
                    gridLocation.Add("column", 0);
                    gridLocation.Add("row", 2);
                    break;
                case 84:
                    gridLocation.Add("column", 3);
                    gridLocation.Add("row", 1);
                    break;
                case 87:
                    gridLocation.Add("column", 6);
                    gridLocation.Add("row", 1);
                    break;
                case 91:
                    gridLocation.Add("column", 9);
                    gridLocation.Add("row", 0);
                    break;
                case 93:
                    gridLocation.Add("column", 7);
                    gridLocation.Add("row", 0);
                    break;
                case 95:
                    gridLocation.Add("column", 5);
                    gridLocation.Add("row", 0);
                    break;
                case 98:
                    gridLocation.Add("column", 2);
                    gridLocation.Add("row", 0);
                    break;
                case 100:
                    gridLocation.Add("column", 0);
                    gridLocation.Add("row", 0);
                    break;
            }

            return gridLocation;
        }
    }
}

File: src/engine/GameEngine.ts
import { Player } from '../domain/Player.js';
import { SpecialMove } from './SpecialMove.js';
import { GameStateEngine } from './GameStateEngine.js';

export class GameEngine {
  players: Player[] = [];
  specialMoves: SpecialMove[] = [];
  currentPlayer: Player | null = null;
  currentState: GameStateEngine = GameStateEngine.InitialGameState;
  currentSpecialMove: SpecialMove | null = null;

  protected randomNumberGenerator: Random;
  protected arrowSpinDuration: number = 0;
  protected numberOfMovesForCurrentPlayer: number = 0;
  protected waitStart: number = 0;

  constructor() {
    this.randomNumberGenerator = new Random();
    this.createListSpecialMoves();
  }

  getNextPlayer(): void {
    if (this.currentPlayer === null) {
      this.currentPlayer = this.players.find(x => x.turnOrder === 1) || null;
    } else {
      const nextTurnOrder = this.currentPlayer.turnOrder + 1;

      this.currentPlayer = nextTurnOrder > this.players.length
        ? this.players.find(x => x.turnOrder === 1) || null
        : this.players.find(x => x.turnOrder === nextTurnOrder) || null;
    }

    this.currentState = GameStateEngine.TurnComplete;
  }

  processArrowEvent(): number {
    this.arrowSpinDuration *= 0.98;

    if (this.randomNumberGenerator.nextDouble() < 0.05 && this.arrowSpinDuration < 360) {
      this.currentState = GameStateEngine.ArrowDelayedEvent;
      this.numberOfMovesForCurrentPlayer = this.calculateNumberOfPlayerMoves(this.arrowSpinDuration);
      this.waitStart = Date.now();
    }

    return this.arrowSpinDuration;
  }

  calculateNumberOfPlayerMoves(arrowSpinAngle: number): number {
    if (arrowSpinAngle >= 0 && arrowSpinAngle <= 60) return 2;
    if (arrowSpinAngle > 60 && arrowSpinAngle <= 120) return 1;
    if (arrowSpinAngle > 120 && arrowSpinAngle <= 180) return 6;
    if (arrowSpinAngle > 180 && arrowSpinAngle <= 240) return 5;
    if (arrowSpinAngle > 240 && arrowSpinAngle <= 300) return 4;
    return 3;
  }

  calculateArrowSpin(): void {
    this.randomNumberGenerator = new Random();
    this.arrowSpinDuration = this.randomNumberGenerator.next(2500) + this.randomNumberGenerator.next(200, 800);
  }

  processArrowDelayEvent(): void {
    if ((Date.now() - this.waitStart) / 1000 > 0.45) {
      this.currentState = GameStateEngine.PlayerEvent;
    }
  }

  processPlayerEvent(): boolean {
    if (!this.currentPlayer) return false;

    if ((Date.now() - this.waitStart) > 100) {
      this.waitStart = Date.now();

      if (!this.currentPlayer.turnInProcess) {
        this.currentPlayer.currentMovesRemaining = this.numberOfMovesForCurrentPlayer;
        this.currentPlayer.turnInProcess = true;
      }

      if (this.currentPlayer.currentTileNumber === 0) {
        this.currentPlayer.currentMovesRemaining--;
        this.currentPlayer.currentTileNumber = 1;
        this.currentPlayer.currentGridColumn = 0;
        this.currentPlayer.currentGridRow = 9;
      } else {
        if (this.currentPlayer.currentTileNumber + this.currentPlayer.currentMovesRemaining > 100) {
          this.currentPlayer.currentMovesRemaining = 0;
        } else {
          this.makePlayerMove();
        }
      }

      if (this.currentPlayer.currentMovesRemaining <= 0) {
        this.currentState = GameStateEngine.GetNextPlayer;
        this.currentPlayer.turnInProcess = false;

        const specialMove = this.isSpecialMove(this.currentPlayer.currentTileNumber);

        if (specialMove) {
          this.currentState = GameStateEngine.PlayerSpecialMoveTransportMoveEvent;
        } else if (this.currentPlayer.currentTileNumber === 100) {
          this.currentState = GameStateEngine.WinnerDeclared;
        }
      }

      return true;
    }

    return false;
  }

  makePlayerMove(): void {
    if (!this.currentPlayer) return;

    if (this.currentPlayer.currentTileNumber % 10 === 0) {
      this.currentPlayer.currentGridRow -= 1;
      this.currentPlayer.currentlyOnAlternateRow = this.currentPlayer.currentGridRow % 2 !== 0;
    } else {
      if (this.currentPlayer.currentlyOnAlternateRow) {
        this.currentPlayer.currentGridColumn -= 1;
      } else {
        this.currentPlayer.currentGridColumn += 1;
      }
    }

    this.currentPlayer.currentTileNumber++;
    this.currentPlayer.currentMovesRemaining--;
  }

  makeSpecialMove(): void {
    if (!this.currentPlayer || !this.currentSpecialMove) return;

    this.currentPlayer.currentGridColumn = this.currentSpecialMove.endingGridColumn;
    this.currentPlayer.currentGridRow = this.currentSpecialMove.endingGridRow;
    this.currentPlayer.currentTileNumber = this.currentSpecialMove.endingTileNumber;
    this.currentPlayer.currentlyOnAlternateRow = this.currentSpecialMove.endingGridRow % 2 === 0;
    this.currentState = GameStateEngine.GetNextPlayer;
    this.currentPlayer.turnInProcess = false;
  }

  isSpecialMove(playerTile: number): boolean {
    const result = this.specialMoves.find(x => x.startingTileNumber === playerTile);

    if (result) {
      this.currentSpecialMove = result;
      if (this.currentPlayer) {
        this.currentPlayer.specialMoveTransportDestination = result.endingTileNumber;
      }
    }

    return result !== undefined;
  }

  createListSpecialMoves(): void {
    this.specialMoves = [];

    const specialMoves: { [key: number]: number } = {
      1: 38, 4: 14, 9: 31, 21: 42,
      28: 84, 51: 67, 71: 91, 80: 100,
      16: 2, 47: 26, 49: 11, 56: 53,
      64: 60, 87: 24, 93: 73, 95: 75,
      98: 78
    };

    for (const [startTile, endTile] of Object.entries(specialMoves)) {
      const startingGridLocation = this.getGridColumnRow(parseInt(startTile));
      const endingGridLocation = this.getGridColumnRow(endTile);

      const sMove = new SpecialMove({
        startingTileNumber: parseInt(startTile),
        startingGridColumn: startingGridLocation.column,
        startingGridRow: startingGridLocation.row,
        endingTileNumber: endTile,
        endingGridColumn: endingGridLocation.column,
        endingGridRow: endingGridLocation.row
      });

      this.specialMoves.push(sMove);
    }
  }

  private getGridColumnRow(tileNumber: number): { column: number; row: number } {
    const gridLocations: { [key: number]: { column: number; row: number } } = {
      1: { column: 0, row: 9 },
      2: { column: 1, row: 9 },
      4: { column: 3, row: 9 },
      9: { column: 8, row: 9 },
      11: { column: 9, row: 8 },
      14: { column: 6, row: 8 },
      16: { column: 4, row: 8 },
      21: { column: 0, row: 7 },
      24: { column: 3, row: 7 },
      26: { column: 5, row: 7 },
      28: { column: 7, row: 7 },
      31: { column: 9, row: 6 },
      38: { column: 2, row: 6 },
      42: { column: 1, row: 5 },
      47: { column: 6, row: 5 },
      49: { column: 8, row: 5 },
      51: { column: 9, row: 4 },
      53: { column: 7, row: 4 },
      56: { column: 4, row: 4 },
      60: { column: 0, row: 4 },
      64: { column: 3, row: 3 },
      67: { column: 6, row: 3 },
      71: { column: 9, row: 2 },
      73: { column: 7, row: 2 },
      75: { column: 5, row: 2 },
      78: { column: 2, row: 2 },
      80: { column: 0, row: 2 },
      84: { column: 3, row: 1 },
      87: { column: 6, row: 1 },
      91: { column: 9, row: 0 },
      93: { column: 7, row: 0 },
      95: { column: 5, row: 0 },
      98: { column: 2, row: 0 },
      100: { column: 0, row: 0 }
    };

    return gridLocations[tileNumber] || { column: 0, row: 0 };
  }
}

// Simple Random class to replace C# Random
class Random {
  next(min: number = 0, max?: number): number {
    if (max === undefined) {
      max = min;
      min = 0;
    }
    return Math.floor(Math.random() * (max - min)) + min;
  }

  nextDouble(): number {
    return Math.random();
  }
}
